git/libgit2/transport/http.go

```diff
@@ -0,0 +1,479 @@
+	}
+
+	t.once.Do(func() {
+		if opts.Context != nil {
+			t.ctx = opts.Context
+			t.logger = ctrl.LoggerFrom(t.ctx,
+				"transportType", "http",
+				"url", opts.TargetURL)
+		}
+	})
+	targetURL := opts.TargetURL
+
+	if targetURL == "" {
+		return nil, fmt.Errorf("repository URL cannot be empty")
+	}
+
+	if len(targetURL) > URLMaxLength {
+		return nil, fmt.Errorf("URL exceeds the max length (%d)", URLMaxLength)
+	}
+
+	var proxyFn func(*http.Request) (*url.URL, error)
+	proxyOpts := opts.ProxyOptions
+	if proxyOpts != nil {
+		switch proxyOpts.Type {
+		case git2go.ProxyTypeNone:
+			proxyFn = nil
+		case git2go.ProxyTypeAuto:
+			proxyFn = http.ProxyFromEnvironment
+		case git2go.ProxyTypeSpecified:
+			parsedUrl, err := url.Parse(proxyOpts.Url)
+			if err != nil {
+				return nil, err
+			}
+			proxyFn = http.ProxyURL(parsedUrl)
+		}
+		t.httpTransport.Proxy = proxyFn
+		t.httpTransport.ProxyConnectHeader = map[string][]string{}
+	} else {
+		t.httpTransport.Proxy = nil
+	}
+
+	client, req, err := createClientRequest(targetURL, action, t.httpTransport, opts.AuthOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	client.CheckRedirect = func(req *http.Request, via []*http.Request) error {
+		if len(via) >= 3 {
+			return fmt.Errorf("too many redirects")
+		}
+
+		// golang will change POST to GET in case of redirects.
+		if len(via) >= 0 && req.Method != via[0].Method {
+			if via[0].URL.Scheme == "https" && req.URL.Scheme == "http" {
+				return fmt.Errorf("downgrade from https to http is not allowed: from %q to %q", via[0].URL.String(), req.URL.String())
+			}
+			if via[0].URL.Host != req.URL.Host {
+				return fmt.Errorf("cross hosts redirects are not allowed: from %s to %s", via[0].URL.Host, req.URL.Host)
+			}
+
+			return http.ErrUseLastResponse
+		}
+
+		// Some Git servers (i.e. Gitlab) only support redirection on the GET operations.
+		// Therefore, on the initial GET operation we update the target URL to include the
+		// new target, so the subsequent actions include the correct target URL.
+		// Example of this is trying to access a Git repository without the .git suffix.
+		if req.Response != nil {
+			if newURL, err := req.Response.Location(); err == nil && newURL != nil {
+				if strings.EqualFold(newURL.Host, req.URL.Host) && strings.EqualFold(newURL.Port(), req.URL.Port()) {
+					opts, _ := getTransportOptions(transportOptionsURL)
+					if opts == nil {
+						opts = &TransportOptions{}
+					}
+
+					opts.TargetURL = trimActionSuffix(newURL.String())
+					AddTransportOptions(transportOptionsURL, *opts)
+
+					// show as info, as this should be visible regardless of the
+					// chosen log-level.
+					t.logger.Info("server responded with redirect",
+						"newUrl", opts.TargetURL, "StatusCode", req.Response.StatusCode)
+				}
+			}
+		}
+
+		return nil
+	}
+
+	stream := newManagedHttpStream(t, req, client)
+	stream.recvReply.Add(1)
+	stream.sendRequestBackground()
```

Comment:
In the previous implementation, this is set only when the request method is POST. I couldn't understand the reason for this change.